### ISSUE TITLE
Only allow uploaded videos to be seen on certain domain names

### ISSUE THREAD

**Issue by 9mido** [Chocobozzz/PeerTube#3773]
Only allow uploaded videos to be seen on certain domain names when the uploaded videos are embedded into the website that has that domain name. Configure the uploaded video to not be seen on peertube publicly, but instead can only be seen on a certain domain name when embedded. 

Vimeo does this. It would be great if peertube could do it as well.

**Comment by Chocobozzz** [comment]
Hello,

A plugin can now do such thing with https://github.com/Chocobozzz/PeerTube/commit/88e12d57ffcbd23be30d36acc94c43a382667092

**Comment by 9mido** [comment]
@Chocobozzz 

Awesome thank you!

**Comment by 9mido** [comment]
@Chocobozzz 

Quick question.

Now that this is included, can something like this be done:

If I have a website that has paid videos, can I embed a private peertube video onto the website's domain only? Can I configure peertube so that nobody can access the paid video on peertube, but can only view it on my domain?

If not, let me know I will make another issue/feature for that. I would love to use peertube for this instead of Vimeo for paid videos on my website.

**Comment by Chocobozzz** [comment]
> If I have a website that has paid videos, can I embed a private peertube video onto the website's domain only? Can I configure peertube so that nobody can access the paid video on peertube, but can only view it on my domain?

Yes, using the `Referer` request header

**Comment by 9mido** [comment]
@Chocobozzz 

Great thanks. I think this is a great step towards moving people off of more centralized platforms. If this is true, I will gladly use Peertube for all of my customer's paid business videos.

**Comment by help4bis** [comment]
@Chocobozzz 
Eh... question.. 
Where do I find these files on my server... I do not have client.js for example.. 

Sorry for the noob question.. 
Thanks
H

**Comment by help4bis** [comment]
Would not mind knowing how this works with eg the embedded link in iframe or if there is another way.. 

I did install the files and now have a 502 error on niginx server... guess a bit more research on my part ... lol.

**Comment by 9mido** [comment]
@help4bis I agree. If you figure it out, please reply with your solution. Maybe post your 502 error and any logs also so others can help.

**Comment by help4bis** [comment]
I get this error... (pls let me know if this is not the wright forum to post this.. happy to move it... ). 

after I change client.js and the download.js back to the originals I checked the server log using the Peertube system etc. this is what is in there. 

error[4/1/2021, 3:31:00 AM] uncaughtException: Cannot use import statement outside a module /var/www/peertube/versions/peertube-v3.1.0/dist/server/controllers/client.js:1 import * as express from 'express' ^^^^^^ SyntaxError: Cannot use import statement outside a module at compileFunction (<anonymous>) at wrapSafe (

**Comment by help4bis** [comment]
@9mido 

Here is some good info on how to use the referrer. 
[https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy#Integration_with_HTML](url)

All I have to now is prevent the server from crashing LOL.. when uploading the two files.... fingers crossed will try tomorrow.. I have to work up the courage first.

**Comment by help4bis** [comment]
@9mido  did you actually get these updates installed and working? I keep getting a 502 bad gateway error... so I wll have to abandon the peertube project.... bummer. 

If you did get it working can you let me know how you got arround the 502 gateway error?  

Thanks
H

**Comment by 9mido** [comment]
@help4bis I did not try it out and I have yet to try peertube out yet either. I wanted this issue to get done so I could have an excuse not to use Vimeo for my website's embeds. Can you confirm if this is directly related to https://github.com/Chocobozzz/PeerTube/commit/88e12d57ffcbd23be30d36acc94c43a382667092 ?

@Chocobozzz Any ideas why @help4bis is receiving a 502 error?

**Comment by help4bis** [comment]
@9mido Confirmed. 
What I did is get each file (apart from the fixtures, as I do not have that directory) and copied the content into the existing files. 

I then did a reboot, which is when the server comes up with a 502... got me stumped...then I figured.. hmm I might have to the tcp tunneling as described in the install section. 

Using Ubuntu flavour. 
==============
When I do a curl on the domain I get a 301, if I do a curl on the ip address I get the proper NGINX message as in ... Thanks you for using NGINX. 

The moment I throw it in a browser I get 502.
sudo nginx -t returns no errors. 
======================
 
 sudo journalctl -fue peertube throws in the log like

0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'start' ]
2 info using npm@6.14.11
3 info using node@v12.22.0
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle peertube@3.1.0~prestart: peertube@3.1.0
6 info lifecycle peertube@3.1.0~start: peertube@3.1.0
7 verbose lifecycle peertube@3.1.0~start: unsafe-perm in lifecycle true
8 verbose lifecycle peertube@3.1.0~start: PATH: /usr/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/var/www/peertube/versions/peertube-v3.1.0/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
9 verbose lifecycle peertube@3.1.0~start: CWD: /var/www/peertube/versions/peertube-v3.1.0
10 silly lifecycle peertube@3.1.0~start: Args: [ '-c', 'node dist/server' ]
11 silly lifecycle peertube@3.1.0~start: Returned: code: 1  signal: null
12 info lifecycle peertube@3.1.0~start: Failed to exec start script
13 verbose stack Error: peertube@3.1.0 start: `node dist/server`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:314:20)
13 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:314:20)
13 verbose stack     at maybeClose (internal/child_process.js:1022:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:287:5)
14 verbose pkgid peertube@3.1.0
15 verbose cwd /var/www/peertube/versions/peertube-v3.1.0
16 verbose Linux 5.4.0-70-generic
17 verbose argv "/usr/bin/node" "/usr/bin/npm" "start"
18 verbose node v12.22.0
19 verbose npm  v6.14.11
20 error code ELIFECYCLE
21 error errno 1
22 error peertube@3.1.0 start: `node dist/server`
22 error Exit status 1
23 error Failed at the peertube@3.1.0 start script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]

Basically no idea... if I roll back all the changes it works just fine... so something has gone astray in my settings when copying it all across...
